bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieWatcher.java

```diff
@@ -138,17 +142,36 @@ private synchronized void processReadOnlyBookiesChanged(Set<BookieSocketAddress>
      *
      * @throws BKException when failed to read bookies
      */
-    public void readBookiesBlocking() throws BKException {
-        this.registrationClient.watchReadOnlyBookies(bookies -> processReadOnlyBookiesChanged(bookies.getValue()));
-        this.registrationClient.watchWritableBookies(bookies -> processWritableBookiesChanged(bookies.getValue()));
+    public void initialBlockingBookieRead() throws BKException {
+        synchronized (this) {
+            if (initialReadonlyBookiesFuture == null
+                && initialWritableBookiesFuture == null) {
+                CompletableFuture<?> writable
+                    = this.registrationClient.watchWritableBookies(
+                            bookies -> processWritableBookiesChanged(bookies.getValue()));
+
+                CompletableFuture<?> readonly
+                    = this.registrationClient.watchReadOnlyBookies(
+                            bookies -> processReadOnlyBookiesChanged(bookies.getValue()));
+                initialWritableBookiesFuture = writable;
```

Comment:
Just being defensive. I want to either assign either both or neither. Also, I need readonly and writable in any case to avoid hitting synchronization errors in findbugs.